editor/test/util/coll_test.clj

```diff
@@ -1584,6 +1710,128 @@
+
+      (testing "Top-level."
+        (doseq [make-coll make-coll-fns]
+          (is (= [{:type :wanted}]
+                 (coll/find-values
+                   #(= :wanted (:type %))
+                   (make-coll {:type :wanted}))))))
+
+      (testing "Nested."
+        (doseq [make-inner-coll make-coll-fns
+                make-middle-coll make-coll-fns
+                make-outer-coll make-coll-fns]
+          (is (= [{:type :wanted}]
+                 (coll/find-values
+                   #(= :wanted (:type %))
+                   (make-outer-coll
+                     (make-middle-coll
+                       (make-inner-coll {:type :wanted}))))))))))
+
+  (testing "Match is returned as-is."
+    (is (= [{:type :wanted
+             :vec [{:type :wanted
+                    :map {:a 1}
+                    :vec [0]}]
+             :map {:type :wanted
+                   :map {:a 1}
+                   :vec [0]}}]
+           (coll/find-values
+             #(= :wanted (:type %))
+             [{:type :wanted
+               :vec [{:type :wanted
+                      :map {:a 1}
+                      :vec [0]}]
+               :map {:type :wanted
+                     :map {:a 1}
+                     :vec [0]}}]))))
+
+  (testing "Over sequence."
+    (is (= [{:type :wanted :index 0}
+            {:type :wanted :index 1}
+            {:type :wanted :index 2}]
+           (coll/find-values
+             #(= :wanted (:type %))
+             [{:type :wanted :index 0}
+              [{:type :wanted :index 1}]
+              {:key {:type :wanted :index 2}}]))))
+
+
+  (testing "As transducer."
+    (is (= [{:type :wanted :index 0}
+            {:type :wanted :index 1}
+            {:type :wanted :index 2}]
+           (into []
+                 (coll/find-values #(= :wanted (:type %)))
+                 [{:type :wanted :index 0}
+                  [{:type :wanted :index 1}]
+                  {:key {:type :wanted :index 2}}])))))
+
+(deftest first-where-test
+  (is (= 2 (coll/first-where even? (range 1 4))))
+  (is (nil? (coll/first-where nil? [:a :b nil :d])))
+  (is (= [:d 4] (coll/first-where (fn [[k _]] (= :d k)) (sorted-map :a 1 :b 2 :c 3 :d 4))))
+  (is (= :e (coll/first-where #(= :e %) (list :a nil :c nil :e))))
+  (is (= "f" (coll/first-where #(= "f" %) (sorted-set "f" "e" "d" "c" "b" "a"))))
+  (is (nil? (coll/first-where nil? nil)))
+  (is (nil? (coll/first-where even? nil)))
+  (is (nil? (coll/first-where even? [])))
+  (is (nil? (coll/first-where even? [1 3 5])))
+
+  (testing "stops calling pred after first true"
+    (let [pred (fn/make-call-logger fn/constantly-true)]
+      (is (= 0 (coll/first-where pred (range 10))))
+      (is (= 1 (count (fn/call-logger-calls pred)))))))
+
+(deftest first-index-where-test
+  (is (= 1 (coll/first-index-where even? (range 1 4))))
+  (is (= 2 (coll/first-index-where nil? [:a :b nil :d])))
+  (is (= 3 (coll/first-index-where #(^[char] Character/isDigit %) "abc123def")))
```

Comment:
BTW you can just do `(first-index-where ^[char] Character/isDigit "abc123def")`

Reply:
Nice!